### PR TITLE
fix #16314

### LINKED ISSUE #16314: gc:arc: lib\pure\os.nim Hint [Performance]
While compiling (on windows) with --gc:arc performance related hints are generated by the compiler

### Example

ding.nim:
```
import os
discard os.paramStr(0)
discard os.parseCmdLine("ding")
```
nim c --gc:arc ding.nim 

### Current Output

The following hints apprear on the console:

```
...\.choosenim\toolchains\nim-1.4.2\lib\pure\os.nim(2764, 64) Hint: passing 'ownArgv[i]' to a sink parameter introduces an implicit copy; if possible, rearrange your program's control flow to prevent it [Performance]
...\.choosenim\toolchains\nim-1.4.2\lib\pure\os.nim(2669, 17) Hint: passing 'a' to a sink parameter introduces an implicit copy; if possible, rearrange your program's control flow to prevent it [Performance]
```

### Expected Output

nothing

### PR DESCRIPTION
**Pull request by ringabout**
fix #16314

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -2587,7 +2587,6 @@ proc parseCmdLine*(c: string): seq[string] {.
   var i = 0
   var a = ""
   while true:
-    setLen(a, 0)
```
Comment (ringabout) [inline lib/pure/os.nim:2590]: Maybe not delete this line to make program more clear?

File: lib/pure/os.nim [hunk 2]
```diff
@@ -2638,7 +2637,7 @@ proc parseCmdLine*(c: string): seq[string] {.
         while i < c.len and c[i] > ' ':
           add(a, c[i])
           inc(i)
-    add(result, a)
+    add(result, move(a))
```
Comment (timotheecour) [inline lib/pure/os.nim:2640]: is that even correct to move something more than once (and should that even compile)?
```nim
proc main(): seq[string] =
  var a = "abc"
  for i in 0..<3: result.add move(a)
echo main()
```
this prints: `@["abc", "", ""]`

**Comment by timotheecour** [comment]
I feel like the pre-existing code is correct, but the hint is bad.
Can't we instead fix this as follows:
```nim
{.push hint[Performance]:off.}
add(result, a)
{.pop.}
```
and making sure this works as intended; this requires a compiler fix regarding treatment of foreignPackageNotes IIRC, given that the pragma is in a foreign package (stdlib) relative to the main project file

refs https://github.com/nim-lang/Nim/pull/16423#issuecomment-748772296

EDIT: => https://github.com/nim-lang/RFCs/issues/317

File: lib/pure/os.nim (hunk 2 above)
Comment (ringabout) [reply lib/pure/os.nim:2640]: It is expected behaviour I think.

Comment (ringabout) [reply lib/pure/os.nim:2640]: Original code
```nim
var a: string
while true:
  a.setLen(0)
  do_something_with(a)
  result.add a
```

After this PR
```nim
var a: string
while true:
  do_something_with(a)
  result.add move(a)
```

Comment (timotheecour) [reply lib/pure/os.nim:2640]: the new code seems worse, at least from performance standpoint:
before PR, a single buffer is reused (`a`), and after each iteration `result.add a` just copies (hopefully eventually via memCopy under the hood) a.len bytes even if a has capacity > a.len.

after PR, a fresh buffer is created at each new iteration, and each iteration can cause a buffer resizes with capacity exceeding the length.

**Comment by n5m** [comment]
What would the right approach have been if `result.add` had taken more than one argument?

For example, if the line in question were instead
```nim
add(result, a, b, c)
```
and we only wanted to suppress the warning on `a`, could we do something like this?
```nim
add(
  result
  {.push hint[Performance]:off.} a {.pop.},
  b,
  c
)
```
I ask because
```nim
{.push hint[Performance]:off.}
add(result, a, b, c)
{.pop.}
```
would _also_ suppress warnings on copying `b` and `c`, which might be undesirable.

Refs https://github.com/nim-lang/Nim/pull/16423#issuecomment-748773523

**Comment by timotheecour** [comment]
> What would the right approach have been if result.add had taken more than one argument?

in this case, this would've worked:
```nim
{.push hint[Performance]:off.}
var a2 = a
{.pop.}
add(result, a2.move, b, c)
```

param pragmas are supported at syntax level:
```nim
{.pragma: foo.}
proc fn(a {.foo.}: int, b: int) = discard
fn(1, 2)
```
but currently not for this pragma

**Comment by timotheecour** [comment]
EDIT: see https://github.com/nim-lang/RFCs/issues/317 for the proposed way to solve this

**Pull request closed without merging**